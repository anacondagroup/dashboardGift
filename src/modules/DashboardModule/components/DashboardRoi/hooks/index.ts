export * from './useRoiDashboardProtector';
export * from './useRoiTable/useRoiTable';
