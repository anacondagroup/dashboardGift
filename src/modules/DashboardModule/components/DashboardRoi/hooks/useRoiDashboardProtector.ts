import { useDispatch, useSelector } from 'react-redux';
import { useEffect } from 'react';

import {
  getIsSfConnectionStateLoaded,
  getSfConnectionState,
} from '../../../../SettingsModule/store/organisation/integrations/workato/integrations/integrations.selectors';
import { ROI_ROUTES } from '../routePaths';
import { fetchSfConnectionState } from '../../../../SettingsModule/store/organisation/integrations/workato/integrations/integrations.actions';

export const useRoiDashboardProtector = (): {
  isLoaded: boolean;
  isSalesforceConnected: boolean;
  dashboard: string;
} => {
  const dispatch = useDispatch();

  const sfConnectionState = useSelector(getSfConnectionState);
  const isLoaded = useSelector(getIsSfConnectionStateLoaded);

  const dashboard = sfConnectionState === 'connected' ? ROI_ROUTES.FUNNEL : ROI_ROUTES.REPORTING;

  useEffect(() => {
    dispatch(fetchSfConnectionState());
  }, [dispatch]);

  return {
    isLoaded,
    isSalesforceConnected: sfConnectionState === 'connected',
    dashboard,
  };
};
