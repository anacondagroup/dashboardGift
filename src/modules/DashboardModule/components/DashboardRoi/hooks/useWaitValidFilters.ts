import { useEffect, useRef } from 'react';
import { TGlobalRoiFilters } from '@alycecom/services';

export const useWaitValidFilters = (filters: TGlobalRoiFilters): boolean => {
  const wasInitialRequestMadeRef = useRef(false);

  useEffect(() => {
    if (filters.teamIds.length) {
      wasInitialRequestMadeRef.current = true;
    }
  }, [filters.teamIds]);

  return wasInitialRequestMadeRef.current ? false : filters.teamIds.length === 0;
};
