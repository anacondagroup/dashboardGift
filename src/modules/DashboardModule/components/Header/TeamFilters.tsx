import React, { memo, useCallback, useMemo } from 'react';
import { useDispatch, useSelector, shallowEqual } from 'react-redux';
import { Link, useRouteMatch } from 'react-router-dom';
import { Grid, MenuItem, Box } from '@mui/material';
import { Icon, DateRangeSelect } from '@alycecom/ui';
import { updateSearch, Auth, User } from '@alycecom/modules';
import { TrackEvent } from '@alycecom/services';
import { useRouting, useUrlQuery, useSetUrlQuery } from '@alycecom/hooks';

import HasPermission from '../../../../hoc/HasPermission/HasPermission';
import { PermissionKeys } from '../../../../constants/permissions.constants';
import { deleteAllGiftsFromTransferSelection } from '../../store/breakdowns/giftTransfer/giftTransfer.actions';
import { makeCanEditTeamById, getTeams } from '../../../../store/teams/teams.selectors';
import SelectFilter from '../../../../components/Dashboard/Header/SelectFilter';

export interface ITeamFiltersProps {
  isLoading?: boolean;
}

const TeamFilters = ({ isLoading = true }: ITeamFiltersProps): JSX.Element => {
  const dispatch = useDispatch();
  const match = useRouteMatch<{ teamId: string }>();

  const teamId = Number(match.params.teamId);
  const go = useRouting();
  const teams = useSelector(getTeams, shallowEqual);
  const canEditTeam = useSelector(useMemo(() => makeCanEditTeamById(match.params.teamId), [match.params.teamId]));

  const { dateRangeFrom = '', dateRangeTo = '' } = useUrlQuery(['dateRangeFrom', 'dateRangeTo']);
  const updateUrl = useSetUrlQuery();

  const onTeamChange = useCallback(
    event => {
      dispatch(deleteAllGiftsFromTransferSelection());
      go(`/teams/${event.teamId}?${updateSearch('', { dateRangeTo, dateRangeFrom })}`);
    },
    [dateRangeFrom, dateRangeTo, dispatch, go],
  );

  const { trackEvent } = TrackEvent.useTrackEvent();
  const userId = useSelector(User.selectors.getUserId);
  const adminId = useSelector(Auth.selectors.getLoginAsAdminId);
  const handleDatesChange = useCallback(
    ({ from, to, preset }) => {
      dispatch(deleteAllGiftsFromTransferSelection());
      updateUrl({ dateRangeFrom: from, dateRangeTo: to });
      trackEvent('Date range - Changed', { page: 'team view', preset, userId, adminId }, { traits: { adminId } });
    },
    [adminId, dispatch, trackEvent, updateUrl, userId],
  );

  return (
    <Grid item container direction="row" justifyContent="flex-end" alignItems="center" xs={5} wrap="nowrap">
      {canEditTeam && (
        <HasPermission permissionKey={PermissionKeys.EditTeams}>
          <Box className="Body-Regular-Center-Link-Bold" pr={2} display="flex" flexDirection="row">
            <Box pr={1}>
              <Icon icon="cog" color="inherit" />
            </Box>
            <Link to={`/settings/teams/${teamId}/settings-and-permissions`}>Edit team</Link>
          </Box>
        </HasPermission>
      )}
      <SelectFilter
        disabled={isLoading}
        label="All teams"
        value={teamId}
        name="teamId"
        onFilterChange={onTeamChange}
        renderItems={() =>
          teams.map(team => (
            <MenuItem key={team.id} value={team.id}>
              {team.name}
            </MenuItem>
          ))
        }
      />
      <Box ml={1}>
        <DateRangeSelect disabled={isLoading} from={dateRangeFrom} to={dateRangeTo} onChange={handleDatesChange} />
      </Box>
    </Grid>
  );
};

export default memo(TeamFilters);
