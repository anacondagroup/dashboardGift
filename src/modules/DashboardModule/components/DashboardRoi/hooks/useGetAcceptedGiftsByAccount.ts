import { TAcceptedGiftByEmailDomain, useGetAcceptedGiftsByAccountQuery } from '@alycecom/services';

import { transformRoiFilters } from '../utils/transformRoiFilters';

import { TDebouncedFilters } from './useRoiTable/useRoiTable';
import { useWaitValidFilters } from './useWaitValidFilters';

export type TUseGetAcceptedGiftsByAccountParams = {
  accountId: string;
  filters: TDebouncedFilters<TAcceptedGiftByEmailDomain>;
};

export const useGetAcceptedGiftsByAccount = ({
  accountId,
  filters,
}: TUseGetAcceptedGiftsByAccountParams): ReturnType<typeof useGetAcceptedGiftsByAccountQuery> & {
  isWaitingForFilters: boolean;
} => {
  const isWaitingForFilters = useWaitValidFilters(filters);
  const queryResult = useGetAcceptedGiftsByAccountQuery(
    {
      filters: transformRoiFilters(filters),
      accountId,
    },
    { skip: isWaitingForFilters },
  );

  return {
    ...queryResult,
    isWaitingForFilters,
  };
};
