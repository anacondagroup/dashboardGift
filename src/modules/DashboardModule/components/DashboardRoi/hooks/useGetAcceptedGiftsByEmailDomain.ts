import { TAcceptedGiftByEmailDomain, useGetAcceptedGiftsByEmailDomainQuery } from '@alycecom/services';

import { transformRoiFilters } from '../utils/transformRoiFilters';

import { TDebouncedFilters } from './useRoiTable/useRoiTable';
import { useWaitValidFilters } from './useWaitValidFilters';

export type TUseGetAcceptedGiftsByEmailDomainParams = {
  emailDomain: string;
  filters: TDebouncedFilters<TAcceptedGiftByEmailDomain>;
};

export const useGetAcceptedGiftsByEmailDomain = ({
  emailDomain,
  filters,
}: TUseGetAcceptedGiftsByEmailDomainParams): ReturnType<typeof useGetAcceptedGiftsByEmailDomainQuery> & {
  isWaitingForFilters: boolean;
} => {
  const isWaitingForFilters = useWaitValidFilters(filters);
  const queryResult = useGetAcceptedGiftsByEmailDomainQuery(
    {
      filters: transformRoiFilters(filters),
      emailDomain,
    },
    { skip: isWaitingForFilters },
  );

  return {
    ...queryResult,
    isWaitingForFilters,
  };
};
