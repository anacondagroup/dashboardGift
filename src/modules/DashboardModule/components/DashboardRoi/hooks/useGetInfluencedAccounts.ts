import { TInfluencedAccounts, useGetInfluencedAccountsQuery } from '@alycecom/services';

import { transformRoiFilters } from '../utils/transformRoiFilters';

import { TDebouncedFilters } from './useRoiTable/useRoiTable';
import { useWaitValidFilters } from './useWaitValidFilters';

export const useGetInfluencedAccounts = (
  filters: TDebouncedFilters<TInfluencedAccounts>,
): ReturnType<typeof useGetInfluencedAccountsQuery> & {
  isWaitingForFilters: boolean;
} => {
  const isWaitingForFilters = useWaitValidFilters(filters);
  const queryResult = useGetInfluencedAccountsQuery(transformRoiFilters(filters), { skip: isWaitingForFilters });

  return {
    ...queryResult,
    isWaitingForFilters,
  };
};
