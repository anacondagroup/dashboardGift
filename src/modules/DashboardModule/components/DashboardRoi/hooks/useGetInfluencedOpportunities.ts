import { TInfluencedOpportunities, useGetInfluencedOpportunitiesQuery } from '@alycecom/services';

import { transformRoiFilters } from '../utils/transformRoiFilters';

import { TDebouncedFilters } from './useRoiTable/useRoiTable';
import { useWaitValidFilters } from './useWaitValidFilters';

export const useGetInfluencedOpportunities = (
  filters: TDebouncedFilters<TInfluencedOpportunities>,
): ReturnType<typeof useGetInfluencedOpportunitiesQuery> & {
  isWaitingForFilters: boolean;
} => {
  const isWaitingForFilters = useWaitValidFilters(filters);
  const queryResult = useGetInfluencedOpportunitiesQuery(transformRoiFilters(filters), { skip: isWaitingForFilters });

  return {
    ...queryResult,
    isWaitingForFilters,
  };
};
