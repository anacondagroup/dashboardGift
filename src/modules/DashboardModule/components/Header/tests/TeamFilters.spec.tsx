import React from 'react';
import { useSelector, shallowEqual } from 'react-redux';
import { useRouteMatch } from 'react-router-dom';
import { shallow } from 'enzyme';
import { create } from 'react-test-renderer';
import { updateSearch, Auth, User } from '@alycecom/modules';
import { useUrlQuery } from '@alycecom/hooks';

import { ISelectFilterProps } from '@alycecom/ui';
import TeamFilters, { ITeamFiltersProps } from '../TeamFilters';
import { getTeams, makeCanEditTeamById } from '../../../../../store/teams/teams.selectors';

const mockedUseRouteMatch = useRouteMatch as jest.Mock;
const mockedUseSelector = useSelector as jest.Mock;
const mockedMakeCanEditTeamById = makeCanEditTeamById as jest.Mock;
const mockedUseUrlQuery = useUrlQuery as jest.Mock;
const mockedUpdateSearch = updateSearch as jest.Mock;

jest.mock('@mui/material', () => ({
  ...jest.requireActual('@mui/material'),
  // eslint-disable-next-line global-require
  makeStyles: require('../../../../../libs/mocks').mockMakeStyles,
  withStyles: jest.fn(() => jest.fn()),
}));
jest.mock('react-redux', () => ({
  connect: jest.fn(() => jest.fn()),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
  shallowEqual: 'shallowEqual',
}));
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn(),
  useRouteMatch: jest.fn(),
  Link: 'Link',
}));
jest.mock('react-router-last-location', () => ({
  useLastLocation: jest.fn(),
}));
jest.mock('@alycecom/hooks', () => ({
  ...jest.requireActual('@alycecom/hooks'),
  useUrlQuery: jest.fn(),
  useRouting: jest.fn(),
  useSetUrlQuery: jest.fn(),
}));
jest.mock('@alycecom/modules', () => ({
  ...jest.requireActual('@alycecom/modules'),
  updateSearch: jest.fn(),
}));
jest.mock('@alycecom/ui', () => ({
  ...jest.requireActual('@alycecom/ui'),
  DateRangeSelect: 'DateRangeSelect',
  Icon: 'Icon',
}));
jest.mock(
  '../../../../../components/Dashboard/Header/SelectFilter',
  () => ({ renderItems, ...props }: Required<ISelectFilterProps>) => <div {...props}>{renderItems()}</div>,
);
jest.mock('../../../../../hoc/HasPermission/HasPermission', () => 'HasPermission');
jest.mock('../../../../../store/teams/teams.selectors');

describe('TeamFilters', () => {
  let props: ITeamFiltersProps;

  beforeEach(() => {
    props = {
      isLoading: true,
    };

    mockedUseRouteMatch.mockReturnValue({ params: { teamId: '1' } });
    mockedMakeCanEditTeamById.mockReset();
    mockedUseUrlQuery.mockReset().mockReturnValue({});
    mockedUpdateSearch.mockReset().mockReturnValue('param=value');
    mockedUseSelector.mockReset().mockReturnValueOnce([
      { id: 1, name: 'Team A' },
      { id: 2, name: 'Horde' },
    ]);
  });

  test('selectors', () => {
    mockedUseSelector.mockReturnValueOnce(false);

    mockedMakeCanEditTeamById.mockReturnValue('makeCanEditTeamByIdResult');
    shallow(<TeamFilters {...props} />);

    expect(mockedMakeCanEditTeamById).toBeCalledWith('1');
    expect(mockedUseUrlQuery).toBeCalled();

    expect(mockedUseSelector.mock.calls).toEqual([
      [getTeams, shallowEqual],
      ['makeCanEditTeamByIdResult'],
      [User.selectors.getUserId],
      [Auth.selectors.getLoginAsAdminId],
    ]);
  });

  it('should be rendered without edit feature', () => {
    mockedUseSelector.mockReturnValueOnce(false);
    // @ts-ignore
    const tree = create(shallow(<TeamFilters {...props} />));

    expect(tree).toMatchSnapshot();
  });

  it('should be rendered with edit feature', () => {
    mockedUseSelector.mockReturnValueOnce(true);

    mockedUseUrlQuery.mockReturnValue({
      dateRangeTo: '2012-12-12',
      dateRangeFrom: '2011-11-11',
    });
    props.isLoading = false;

    // @ts-ignore
    const tree = create(shallow(<TeamFilters {...props} />));

    expect(tree).toMatchSnapshot();
  });
});
