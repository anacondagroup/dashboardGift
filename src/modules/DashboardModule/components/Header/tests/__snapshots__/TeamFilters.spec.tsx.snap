// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamFilters should be rendered with edit feature 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-wrap-xs-nowrap MuiGrid-grid-xs-5 css-ufwsty-MuiGrid-root"
>
  <HasPermission
    permissionKey="edit-teams"
  >
    <div
      className="Body-Regular-Center-Link-Bold MuiBox-root css-1d3mru"
    >
      <div
        className="MuiBox-root css-111wug5"
      >
        <Icon
          color="inherit"
          icon="cog"
        />
      </div>
      <Link
        to="/settings/teams/1/settings-and-permissions"
      >
        Edit team
      </Link>
    </div>
  </HasPermission>
  <div
    disabled={false}
    label="All teams"
    name="teamId"
    onFilterChange={[Function]}
    value={1}
  >
    <li
      className="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="menuitem"
      tabIndex={-1}
      value={1}
    >
      Team A
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </li>
    <li
      className="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="menuitem"
      tabIndex={-1}
      value={2}
    >
      Horde
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </li>
  </div>
  <div
    className="MuiBox-root css-1isemmb"
  >
    <DateRangeSelect
      disabled={false}
      from="2011-11-11"
      onChange={[Function]}
      to="2012-12-12"
    />
  </div>
</div>
`;

exports[`TeamFilters should be rendered without edit feature 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-wrap-xs-nowrap MuiGrid-grid-xs-5 css-ufwsty-MuiGrid-root"
>
  <div
    disabled={true}
    label="All teams"
    name="teamId"
    onFilterChange={[Function]}
    value={1}
  >
    <li
      className="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="menuitem"
      tabIndex={-1}
      value={1}
    >
      Team A
    </li>
    <li
      className="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="menuitem"
      tabIndex={-1}
      value={2}
    >
      Horde
    </li>
  </div>
  <div
    className="MuiBox-root css-1isemmb"
  >
    <DateRangeSelect
      disabled={true}
      from=""
      onChange={[Function]}
      to=""
    />
  </div>
</div>
`;
